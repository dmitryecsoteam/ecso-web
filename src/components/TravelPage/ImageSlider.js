import React from 'react';
import loadImage from '../../utils/loadImage';
import '@babel/polyfill';

export default class ImageSlider extends React.Component {

    // Put first slide in initial state, so it could be displayed as soon as it would be loaded
    // Then load all next slides in componentDidMount
    state = {
        images: [`/images/${this.props.name}-${this.props.country}/slide1.jpg`],
        index: 0
    }

    // ref to ImageSlider div container. To control css classes and animation
    divRef = React.createRef();

    // Interval property for setInterval and clearInterval
    autoSlide = 0;

    async componentDidMount() {

        // add animation
        this.divRef.current.classList.add('image-slider__animation');

        const images = this.state.images.slice();
        let i = 2;
        let loadNextImage = true;
        while (loadNextImage) {

            // load next image
            loadNextImage = await loadImage(`/images/${this.props.name}-${this.props.country}/slide${i}.jpg`);

            // TRUE - image exists and loaded in cache, add this URL to images array and increment counter
            // FALSE - image doesn't exist, leave while loop
            if (loadNextImage) {
                images.push(`/images/${this.props.name}-${this.props.country}/slide${i}.jpg`);
                i++;
            }
        }

        // update state to rerender component
        this.setState({
            images
        });

        if (images.length > 1) {
            // Set interval to auto-slide when there are more than one image
            this.autoSlide = setInterval(this.slideRight, this.props.interval);
        }
    }

    componentDidUpdate() {

        if (this.divRef.current) {

            // HTMLElement's property reference triggers DOM reflow
            // without this animation wouldn't work
            // https://css-tricks.com/restart-css-animation/#article-header-id-0
            this.divRef.current.offsetWidth;

            // rerun animation on new image
            this.divRef.current.classList.add('image-slider__animation');

        }
    }

    componentWillUnmount() {
        clearInterval(this.autoSlide);
    }

    slideAndResetInterval = (slide, id) => {

        // remove auto-slide interval
        clearInterval(this.autoSlide);

        // change background image
        slide(id);

        // set new auto-slide interval
        this.autoSlide = setInterval(this.slideRight, this.props.interval);
    }

    slideTo = (index) => {

        // remove animation
        this.divRef.current.classList.remove('image-slider__animation');

        this.setState({
            index
        });
    }

    slideRight = () => {

        // rerender with next background image from images array
        // if current background image is the last in array of images, than move to the start
        if (this.state.index === this.state.images.length - 1) this.slideTo(0);
        else this.slideTo(this.state.index + 1);

    }

    slideLeft = () => {

        // rerender with previous background image from images array
        // if current background image is the first one, than move to the last one
        if (this.state.index === 0) this.slideTo(this.state.images.length - 1);
        else this.slideTo(this.state.index - 1);

    }

    onLeftClick = () => {

        // slide to previous image
        this.slideAndResetInterval(this.slideLeft);

    }

    onRightClick = () => {

        // slide to next image
        this.slideAndResetInterval(this.slideRight);

    }

    onDotClick = (event) => {

        // slide to choosed image
        this.slideAndResetInterval(this.slideTo, +event.target.dataset.id);

    }

    render() {

        const { index } = this.state;

        const { images } = this.state;

        const dots = [];
        for (let i = 0; i < images.length; i++) {
            let className = "image-slider__dot";
            if (i === index) {
                className += " image-slider__dot--active"
            }
            dots.push(<span key={i} data-id={i} className={className} onClick={this.onDotClick}></span>)
        }

        // const containerClassNames = classNames('image-slider__container',
        //     {
        //         'image-slider__container--empty': images.length === 0
        //     }
        // );

        return (
            <div>
                {images.length !== 0 && <div className="image-slider__container">
                    <div
                        ref={this.divRef}
                        className="image-slider__current-slide"
                        style={{ backgroundImage: `url(${images[index]})` }}
                    >
                    </div>
                    <div className="image-slider__button image-slider__button--left" onClick={this.onLeftClick}>&#10094;</div>
                    <div className="image-slider__button image-slider__button--right" onClick={this.onRightClick}>&#10095;</div>
                    <div className="image-slider__dots-container">{dots}</div>
                </div>}
            </div>
        );
    }
}