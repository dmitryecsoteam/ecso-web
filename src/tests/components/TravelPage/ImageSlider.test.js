import React from 'react';
import { mount } from 'enzyme';
import ImageSlider from '../../../components/TravelPage/ImageSlider';

jest.mock('../../../utils/loadImage.js')

// puts resolve stub function in callback queue
// when this function is called, it waits untill async componentDidMount completes
const waitForAsync = () => new Promise(resolve => setImmediate(resolve));

let wrapper;
beforeEach(async () => {

    wrapper = mount(<ImageSlider
        interval={5000}
        name='Name'
        country='Country'
    />);

    await waitForAsync();

    // re-render after componentDidMount completes
    wrapper.update();
});

test('should render empty ImageSlider', async () => {
    const wrapper = mount(<ImageSlider
        interval={5000}
        name='No'
        country='Images'
    />);

    await waitForAsync();

    // re-render after componentDidMount completes
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state('images')).toEqual([]);
});

test('should render ImageSlider with three images', () => {

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state('images')).toEqual(["/images/Name-Country/slide1.jpg", "/images/Name-Country/slide2.jpg", "/images/Name-Country/slide3.jpg"]);

    // Initial index must be 0
    expect(wrapper.state('index')).toBe(0);
});

test('should rotate image right', () => {

    // 1. Right click
    wrapper.find('.image-slider__button--right').simulate('click');
    wrapper.update();

    // Index should change to 1
    // and bakground image in style should change to slide2
    expect(wrapper.state('index')).toBe(1);
    expect(wrapper.find('.image-slider__current-slide').prop('style')).toEqual({ backgroundImage: 'url(/images/Name-Country/slide2.jpg)' });


    // 2. Right click again
    wrapper.find('.image-slider__button--right').simulate('click');
    wrapper.update();

    // Index should change to 2
    // and bakground image in style should change to slide3
    expect(wrapper.state('index')).toBe(2);
    expect(wrapper.find('.image-slider__current-slide').prop('style')).toEqual({ backgroundImage: 'url(/images/Name-Country/slide3.jpg)' });

    // 3. Right click again
    wrapper.find('.image-slider__button--right').simulate('click');
    wrapper.update();

    // Index should change to 0
    // and bakground image in style should change to slide1
    expect(wrapper.state('index')).toBe(0);
    expect(wrapper.find('.image-slider__current-slide').prop('style')).toEqual({ backgroundImage: 'url(/images/Name-Country/slide1.jpg)' });
});

test('should rotate image left', () => {

    // 1. Left click
    wrapper.find('.image-slider__button--left').simulate('click');
    wrapper.update();

    // Index should change to 2
    // and bakground image in style should change to slide3
    expect(wrapper.state('index')).toBe(2);
    expect(wrapper.find('.image-slider__current-slide').prop('style')).toEqual({ backgroundImage: 'url(/images/Name-Country/slide3.jpg)' });


    // 2. Left click
    wrapper.find('.image-slider__button--left').simulate('click');
    wrapper.update();

    // Index should change to 1
    // and bakground image in style should change to slide2
    expect(wrapper.state('index')).toBe(1);
    expect(wrapper.find('.image-slider__current-slide').prop('style')).toEqual({ backgroundImage: 'url(/images/Name-Country/slide2.jpg)' });


    // 3. Left click
    wrapper.find('.image-slider__button--left').simulate('click');
    wrapper.update();

    // Index should change to 0
    // and bakground image in style should change to slide1
    expect(wrapper.state('index')).toBe(0);
    expect(wrapper.find('.image-slider__current-slide').prop('style')).toEqual({ backgroundImage: 'url(/images/Name-Country/slide1.jpg)' });
});

test('should change image on dot click', async () => {

    // We are currently on slide1
    expect(wrapper.find('.image-slider__current-slide').prop('style')).toEqual({ backgroundImage: 'url(/images/Name-Country/slide1.jpg)' });

    // Click on last dot
    const dot = wrapper.find('.image-slider__dot').last();
    dot.simulate('click', { target: { dataset: { id: dot.prop('data-id') } }});
    wrapper.update();

    // Index should change to 2
    // and bakground image in style should change to slide3
    expect(wrapper.state('index')).toBe(2);
    expect(wrapper.find('.image-slider__current-slide').prop('style')).toEqual({ backgroundImage: 'url(/images/Name-Country/slide3.jpg)' });
});

test('should set correct interval', async () => {

    jest.useFakeTimers();

    const interval = 15000;

    const wrapper = mount(<ImageSlider
        interval={interval}
        name='No'
        country='Images'
    />);

    await waitForAsync();

    // should call setInterval on component mount
    expect(setInterval).toHaveBeenCalledWith(expect.any(Function), interval);

    
    // change image
    wrapper.find('.image-slider__button--right').simulate('click');
    // should reset interval
    expect(setInterval).toHaveBeenCalledTimes(2);
    expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), interval);
});