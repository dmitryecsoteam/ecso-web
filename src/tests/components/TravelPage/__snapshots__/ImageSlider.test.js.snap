// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ImageSlider with three images 1`] = `
<ImageSlider
  country="Country"
  interval={5000}
  name="Name"
>
  <div>
    <div
      className="image-slider__container"
    >
      <div
        className="image-slider__current-slide"
        style={
          Object {
            "backgroundImage": "url(/images/Name-Country/slide1.jpg)",
          }
        }
      />
      <div
        className="image-slider__button image-slider__button--left"
        onClick={[Function]}
      >
        ❮
      </div>
      <div
        className="image-slider__button image-slider__button--right"
        onClick={[Function]}
      >
        ❯
      </div>
      <div
        className="image-slider__dots-container"
      >
        <span
          className="image-slider__dot image-slider__dot--active"
          data-id={0}
          key="0"
          onClick={[Function]}
        />
        <span
          className="image-slider__dot"
          data-id={1}
          key="1"
          onClick={[Function]}
        />
        <span
          className="image-slider__dot"
          data-id={2}
          key="2"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
</ImageSlider>
`;

exports[`should render empty ImageSlider 1`] = `
<ImageSlider
  country="Images"
  interval={5000}
  name="No"
>
  <div />
</ImageSlider>
`;
